Alex and Happy Part 1/article.md

```diff
@@ -0,0 +1,728 @@
+
+```text
+                 ┌─┐
+                 ▼ │ [a-z]
+                ┌──┴┐
+                │┌─┐│
+        ┌──────►││4││◄──────┬───────┐
+        │       │└─┘│       │       │
+        │       └───┘       │       │
+        │         ▲         │       │
+        │ [a-z]   │ [a-z]   │       │
+        │         │         │ [a-z] │
+      ┌─┴─┐     ┌─┴─┐     ┌─┴─┐     │
+      │   │ i   │┌─┐│ n   │┌─┐│     │
+─────►│ 0 ├────►││1│├────►││2││     │
+      │   │     │└─┘│     │└─┘│     │
+      └───┘     └─┬─┘     └───┘     │
+                  │                 │
+                  │f                │
+                  ▼                 │
+                ┌───┐               │
+                │┌─┐│ [a-z]         │
+                ││3│├───────────────┘
+                │└─┘│
+                └───┘
+```
+
+To simplify matters, on ambiguous transitions such as `i` and `[a-z]`, assume that the one with a single letter takes precedence.
+
+State 0 is the initial state where the automaton begins. Suppose that the automaton can stop at states 1, 2, 3, and 4.
+
+If we want to check the expression "int", we will follow this path: 0 → 1 → 2 → 4.
+
+Now, we need to attribute meanings to each state so that this automaton can be useful, and we will do so like this:
+* 0: Error.
+* 1: Identifier.
+* 2: In.
+* 3: If.
+* 4: Identifier.
+
+Since we stopped at 4, the result is that "int" is an identifier, so we get a token that is `Identifier "int"`.
+The resulting token would be 'In' if we stopped at 2 instead (for the expression "in").
+
+
+What if we consumed no input and stopped at state 0, or the input was something like "Foo123" where there aren't valid transitions?
+In such cases, we halt and indicate a lexical error.
+
+We won't write our lexical analyzer from scratch. Instead, we will use Alex to create a list of tokens, which Happy will parse.
+
+### Our first lexer
+
+Begin by creating a new file: `src/Lexer.x`.
+For now, let's add a "skeleton" for the file, which should look like the following:
+
+```alex
+{
+module Lexer
+  ( -- * Invoking Alex
+    Alex
+  , AlexPosn (..)
+  , alexGetInput
+  , alexError
+  , runAlex
+  , alexMonadScan
+
+  , Range (..)
+  , Token (..)
+  , TokenType (..)
+  ) where
+}
+
+%wrapper "monadUserState-bytestring"
+
+tokens :-
+
+<0> $white+ ;
+
+{
+data AlexUserState = AlexUserState
+  {
+  }
+
+alexInitUserState :: AlexUserState
+alexInitUserState = AlexUserState
+
+alexEOF :: Alex Token
+alexEOF = do
+  (pos, _, _, _) <- alexGetInput
+  pure $ Token EOF (Range pos pos)
+
+data Range = Range
+  { start :: AlexPosn
+  , stop :: AlexPosn
+  } deriving (Eq, Show)
+
+data Token = Token
+  { tType :: TokenType
+  , tRange :: Range
+  } deriving (Eq, Show)
+
+data TokenType
```

Comment:
I've renamed it, as well as in the GitHub gists. I didn't update part 2 (about Happy) yet, though, but I will do so soon.

I actually noticed that in the Happy tutorial, I use the version of the lexer without the tutorial (i.e., strings are still dumb, and there is `Integer` instead of `Number`), so I will also update it to include these solutions.